### ISSUE TITLE
HttpListener Stack over flow

### ISSUE THREAD

**Issue by DebugST** [dotnet/runtime#73674]
### Description

This is my code:

```c#
var hl = new HttpListener();
hl.Prefixes.Add("...");
hl.Start();
hl.BeginGetContext(HttpCallBack, hl);

private void HttpCallBack(IAsyncResult ir) {
    HttpListener hl = ir.AsyncState as HttpListener;
    HttpListenerContext context = hl.EndGetContext(ir);
    hl.BeginGetContext(HttpCallBack, hl);
    //some other code
}
```
But after some time an exception occurs:
```
Stack overflow.
at Interop+Sys.Accept(System.Runtime.InteropServices.SafeHandle, Byte*, Int32*, IntPtr*)
at System.Net.Sockets.SocketPal.TryCompleteAccept(System.Net.Sockets.SafeSocketHandle, Byte[], Int32 ByRef, IntPtr ByRef, System.Net.Sockets.SocketError ByRef)
at System.Net.Sockets.SocketAsyncContext.AcceptAsync(Byte[], Int32 ByRef, IntPtr ByRef, System.Action`4<IntPtr,Byte[],Int32,System.Net.Sockets.SocketError>)
at System.Net.Sockets.SocketAsyncEventArgs.DoOperationAccept(System.Net.Sockets.Socket, System.Net.Sockets.SafeSocketHandle, System.Net.Sockets.SafeSocketHandle)
at System.Net.Sockets.Socket.AcceptAsync(System.Net.Sockets.SocketAsyncEventArgs)
at System.Net.HttpEndPointListener.Accept(System.Net.Sockets.SocketAsyncEventArgs)
```
Then I found the code for `HttpEndPointListener`:[https://github.com/dotnet/runtime/blob/eeeeb7a75c338b0d994ccd7140d676a612990ebc/src/libraries/System.Net.HttpListener/src/System/Net/Managed/HttpEndPointListener.cs#L100](https://github.com/dotnet/runtime/blob/eeeeb7a75c338b0d994ccd7140d676a612990ebc/src/libraries/System.Net.HttpListener/src/System/Net/Managed/HttpEndPointListener.cs#L100)
This is the Code:

```c#
class HttpEndPointListener{
    public HttpEndPointListener(HttpListener listener, IPAddress addr, int port, bool secure)
    {
        // some other code
        _endpoint = new IPEndPoint(addr, port);
        _socket = new Socket(addr.AddressFamily, SocketType.Stream, ProtocolType.Tcp);
        _socket.Bind(_endpoint);
        _socket.Listen(500);

        SocketAsyncEventArgs args = new SocketAsyncEventArgs();
        args.UserToken = this;
        args.Completed += OnAccept;
        Accept(args);
        // some other code
    }

    private void Accept(SocketAsyncEventArgs e)
    {
        e.AcceptSocket = null;
        bool asyn;
        try
        {
            asyn = _socket.AcceptAsync(e);
        }
        catch (ObjectDisposedException)
        {
            return;
        }
        // if asyn always `false`. This will be an infinite recursion.
        // Because my code runs on a local area network, the network latency is low.
        if (!asyn)
        {
            ProcessAccept(e);
        }
    }

    private static void ProcessAccept(SocketAsyncEventArgs args)
    {
        HttpEndPointListener epl = (HttpEndPointListener)args.UserToken!;

        Socket? accepted = args.SocketError == SocketError.Success ? args.AcceptSocket : null;
        // If there is already a completed connection in the current protocol stack.
        // Then the `asyn` will get the `false` in `Accept`.
        // And then call `ProcessAccept` in Accept.
        epl.Accept(args);
        // some other code.
    }
}
```

### Reproduction Steps

Place the service program in the local area network, or this machine. Then a large number of connections are established with `HttpListener`.

### Expected behavior

```
Stack overflow.
at Interop+Sys.Accept(System.Runtime.InteropServices.SafeHandle, Byte*, Int32*, IntPtr*)
at System.Net.Sockets.SocketPal.TryCompleteAccept(System.Net.Sockets.SafeSocketHandle, Byte[], Int32 ByRef, IntPtr ByRef, System.Net.Sockets.SocketError ByRef)
at System.Net.Sockets.SocketAsyncContext.AcceptAsync(Byte[], Int32 ByRef, IntPtr ByRef, System.Action`4<IntPtr,Byte[],Int32,System.Net.Sockets.SocketError>)
at System.Net.Sockets.SocketAsyncEventArgs.DoOperationAccept(System.Net.Sockets.Socket, System.Net.Sockets.SafeSocketHandle, System.Net.Sockets.SafeSocketHandle)
at System.Net.Sockets.Socket.AcceptAsync(System.Net.Sockets.SocketAsyncEventArgs)
at System.Net.HttpEndPointListener.Accept(System.Net.Sockets.SocketAsyncEventArgs)
```

### Actual behavior

Stack overflow.

### Regression?

The code in this page is also problematic. When I saw it before, I thought it was just a `Demo`. I didn't expect that there was a similar way of writing in `HttpEndPointListener`.
[https://docs.microsoft.com/en-us/dotnet/api/system.net.sockets.socketasynceventargs?redirectedfrom=MSDN&view=net-6.0](https://docs.microsoft.com/en-us/dotnet/api/system.net.sockets.socketasynceventargs?redirectedfrom=MSDN&view=net-6.0)

### Known Workarounds

```c#
public delegate void IOProcessHandler(SocketAsyncEventArgs e);

public static class IOProcessPool
{
    private static ManualResetEvent m_mre;
    private static Stack<IOHandlerInfo> m_stack_idle;
    private static Queue<IOHandlerInfo> m_queue_work;

    static IOProcessPool() {
        IOHandlerInfo hi = null;
        m_mre = new ManualResetEvent(false);
        m_stack_idle = new Stack<IOHandlerInfo>();
        m_queue_work = new Queue<IOHandlerInfo>();
        new Thread(() => {
            while (true) {
                hi = null;
                lock (m_queue_work) {
                    if (m_queue_work.Count != 0) hi = m_queue_work.Dequeue();
                }
                if (hi == null) {
                    m_mre.WaitOne();
                    m_mre.Reset();
                    continue;
                }
                hi.Handler(hi.Args);
                IOProcessPool.PushHandler(hi);
            }
        }) { IsBackground = true }.Start();
    }

    private static IOHandlerInfo PopHandler(IOProcessHandler handler, SocketAsyncEventArgs args) {
        IOHandlerInfo hi = null;
        lock (m_stack_idle) {
            if (m_stack_idle.Count != 0) hi = m_stack_idle.Pop();
        }
        if (hi == null) {
            hi = new IOHandlerInfo(handler, args);
        } else {
            hi.Handler = handler;
            hi.Args = args;
        }
        return hi;
    }

    private static void PushHandler(IOHandlerInfo hi) {
        lock (m_stack_idle) m_stack_idle.Push(hi);
    }

    public static void QueueWork(IOProcessHandler handler, SocketAsyncEventArgs args) {
        lock (m_queue_work) {
            m_queue_work.Enqueue(IOProcessPool.PopHandler(handler, args));
        }
        m_mre.Set();
    }

    private class IOHandlerInfo
    {
        private IOProcessHandler _Handler;

        public IOProcessHandler Handler {
            get { return _Handler; }
            set { _Handler = value; }
        }

        private SocketAsyncEventArgs _Args;

        public SocketAsyncEventArgs Args {
            get { return _Args; }
            set { _Args = value; }
        }

        public IOHandlerInfo(IOProcessHandler handler, SocketAsyncEventArgs args) {
            this._Handler = handler;
            this._Args = args;
        }
    }
}
```
And In the `HttpEndPointListener.Accept`:

```c#
if (!asyn)
{
    IOProcessPool.QueueWork(ProcessAccept, e);
}
```
Of course, the above is just an idea, I believe you have a better way, after all, you are more professional.

**Comment by DebugST** [comment]
```
Stack overflow.
at Interop+Sys.Accept(System.Runtime.InteropServices.SafeHandle, Byte*, Int32*, IntPtr*)
at System.Net.Sockets.SocketPal.TryCompleteAccept(System.Net.Sockets.SafeSocketHandle, Byte[], Int32 ByRef, IntPtr ByRef, System.Net.Sockets.SocketError ByRef)
at System.Net.Sockets.SocketAsyncContext.AcceptAsync(Byte[], Int32 ByRef, IntPtr ByRef, System.Action`4<IntPtr,Byte[],Int32,System.Net.Sockets.SocketError>)
at System.Net.Sockets.SocketAsyncEventArgs.DoOperationAccept(System.Net.Sockets.Socket, System.Net.Sockets.SafeSocketHandle, System.Net.Sockets.SafeSocketHandle)
at System.Net.Sockets.Socket.AcceptAsync(System.Net.Sockets.SocketAsyncEventArgs)
at System.Net.HttpEndPointListener.Accept(System.Net.Sockets.SocketAsyncEventArgs)
at System.Net.HttpEndPointListener.ProcessAccept(System.Net.Sockets.SocketAsyncEventArgs)
at System.Net.HttpEndPointListener.ProcessAccept(System.Net.Sockets.SocketAsyncEventArgs)
at System.Net.HttpEndPointListener.ProcessAccept(System.Net.Sockets.SocketAsyncEventArgs)
at System.Net.HttpEndPointListener.ProcessAccept(System.Net.Sockets.SocketAsyncEventArgs)
...
...
at System.Net.HttpEndPointListener.ProcessAccept(System.Net.Sockets.SocketAsyncEventArgs)
```

**Comment by teo-tsirpanis** [comment]
This can happen if `BeginGetContext` completes synchronously, which you can check using `IAsyncResult.CompletedSynchronously`.

Have you thought using `await hl.GetContextAsync()` in a loop instead? This is not susceptible to stack overflows and will make your code simpler.

**Comment by DebugST** [comment]
> This can happen if `BeginGetContext` completes synchronously, which you can check using `IAsyncResult.CompletedSynchronously`.
> 
> Have you thought using `await hl.GetContextAsync()` in a loop instead? This is not susceptible to stack overflows and will make your code simpler.

I found the code in this link:
[https://github.com/dotnet/runtime/blob/eeeeb7a75c338b0d994ccd7140d676a612990ebc/src/libraries/System.Net.HttpListener/src/System/Net/HttpListener.cs#L259](https://github.com/dotnet/runtime/blob/eeeeb7a75c338b0d994ccd7140d676a612990ebc/src/libraries/System.Net.HttpListener/src/System/Net/HttpListener.cs#L259)

```
public Task<HttpListenerContext> GetContextAsync()
{
    return Task.Factory.FromAsync(
        (callback, state) => ((HttpListener)state!).BeginGetContext(callback, state),
        iar => ((HttpListener)iar!.AsyncState!).EndGetContext(iar),
        this);
}
```
so the `BeginGetContext` the same as `GetContextAsync`???
And ...... the exception is:
```
Stack overflow.
at Interop+Sys.Accept(System.Runtime.InteropServices.SafeHandle, Byte*, Int32*, IntPtr*)
at System.Net.Sockets.SocketPal.TryCompleteAccept(System.Net.Sockets.SafeSocketHandle, Byte[], Int32 ByRef, IntPtr ByRef, System.Net.Sockets.SocketError ByRef)
at System.Net.Sockets.SocketAsyncContext.AcceptAsync(Byte[], Int32 ByRef, IntPtr ByRef, System.Action`4<IntPtr,Byte[],Int32,System.Net.Sockets.SocketError>)
at System.Net.Sockets.SocketAsyncEventArgs.DoOperationAccept(System.Net.Sockets.Socket, System.Net.Sockets.SafeSocketHandle, System.Net.Sockets.SafeSocketHandle)
at System.Net.Sockets.Socket.AcceptAsync(System.Net.Sockets.SocketAsyncEventArgs)
at System.Net.HttpEndPointListener.Accept(System.Net.Sockets.SocketAsyncEventArgs)
at System.Net.HttpEndPointListener.ProcessAccept(System.Net.Sockets.SocketAsyncEventArgs)
at System.Net.HttpEndPointListener.ProcessAccept(System.Net.Sockets.SocketAsyncEventArgs)
at System.Net.HttpEndPointListener.ProcessAccept(System.Net.Sockets.SocketAsyncEventArgs)
at System.Net.HttpEndPointListener.ProcessAccept(System.Net.Sockets.SocketAsyncEventArgs)
...
...
at System.Net.HttpEndPointListener.ProcessAccept(System.Net.Sockets.SocketAsyncEventArgs)
```
`at System.Net.HttpEndPointListener.ProcessAccept(System.Net.Sockets.SocketAsyncEventArgs)`
`ProcessAccept`....not `BeginGetContext`
😃😃😃😃

**Comment by teo-tsirpanis** [comment]
> so the `BeginGetContext` the same as `GetContextAsync`???
No, `GetContextAsync` does call `BeginGetContext` but in a safe way that does not cause stack overflows.

And in any case using it together with the `await` keyword would _greatly_ simplify your code.

**Comment by DebugST** [comment]
> > so the `BeginGetContext` the same as `GetContextAsync`???
> > No, `GetContextAsync` does call `BeginGetContext` but in a safe way that does not cause stack overflows.
> 
> And in any case using it together with the `await` keyword would _greatly_ simplify your code.

Thank you very much for your patient answer. I don't mean there is no difference between `BeginGetContext` and `GetContextAsync`, but that `BeginGetContext` is still called inside `GetContextAsync`.
What caused the stack overflow was not how my code should be written, but the stack overflow caused by the `ProcessAccept` of `HttpEndPointListener`.
The `HttpEndPointListener` is used in the `HttpListener` to listen on the port, and the `HttpEndPointListener.ProcessAccept()` contains a recursive call.
With this link you can view line 100 of the `HttpEndPointListener`.
[https://github.com/dotnet/runtime/blob/eeeeb7a75c338b0d994ccd7140d676a612990ebc/src/libraries/System.Net.HttpListener/src/System/Net/Managed/HttpEndPointListener.cs#L100](https://github.com/dotnet/runtime/blob/eeeeb7a75c338b0d994ccd7140d676a612990ebc/src/libraries/System.Net.HttpListener/src/System/Net/Managed/HttpEndPointListener.cs#L100)

Below is the key code：
``` cs
class HttpEndPointListener{
    // some other code
    private void Accept(SocketAsyncEventArgs e)
    {
        e.AcceptSocket = null;
        bool asyn;
        try
        {
            asyn = _socket.AcceptAsync(e);
        }
        catch (ObjectDisposedException)
        {
            return;
        }
        // if asyn always `false`. This will be an infinite recursion.
        // Because my code runs on a local area network, the network latency is low.
        if (!asyn)
        {
            ProcessAccept(e);
        }
    }

    private static void ProcessAccept(SocketAsyncEventArgs args)
    {
        HttpEndPointListener epl = (HttpEndPointListener)args.UserToken!;

        Socket? accepted = args.SocketError == SocketError.Success ? args.AcceptSocket : null;
        // If there is already a completed connection in the current protocol stack.
        // Then the `asyn` will get the `false` in `epl.Accept()`.
        // And then call `ProcessAccept` in `epl.Accept()`.
        epl.Accept(args);
        // some other code.
    }
}
```
``` cs
ProcessAccept() -> call epl.Accept() -> if asyn = false, Call ProcessAccept() directly ...
```

**Comment by teo-tsirpanis** [comment]
Ohh, I now understand what happened. `AcceptAsync` returned false which means it completed synchronously, so it processes the connection and tries to asynchronously accept again. It was getting too many connections, they all completed synchronously and boom, stack overflow.

**Comment by DebugST** [comment]
> Ohh, I now understand what happened. `AcceptAsync` returned false which means it completed synchronously, so it processes the connection and tries to asynchronously accept again. It was getting too many connections, they all completed synchronously and boom, stack overflow.

the same code also in this link:
[https://docs.microsoft.com/en-us/dotnet/api/system.net.sockets.socketasynceventargs?redirectedfrom=MSDN&view=net-6.0](https://docs.microsoft.com/en-us/dotnet/api/system.net.sockets.socketasynceventargs?redirectedfrom=MSDN&view=net-6.0)

I found this code a long time ago with stack overflow, but I thought it was just a `Demo` telling the reader how to use `SocketAsyncEventargs`. So I didn't write the code according to the case, but added a class to process it, which is the `IOProcessPool` I mentioned at the top.

So when I saw that the exception was a stack overflow, and saw the keyword `AcceptAsync`, I probably guessed what was going on. As a result, similar writing was found in `HttpEndPointListener`.
😁😁😁
